Mathlib/Data/BitVec.lean

```diff
@@ -21,6 +22,26 @@ namespace BitVec
 
 variable {w : Nat}
 
+-- TODO: move to the Lean4 repository.
+open Fin.CommRing in
+theorem ofFin_intCast (z : ℤ) : ofFin (z : Fin (2^w)) = ↑z := by
+  cases w
+  case zero =>
+    simp only [eq_nil]
+  case succ w =>
+    apply BitVec.eq_of_toInt_eq
+    rw [toInt_ofFin, Fin.val_intCast, Int.natCast_pow, Nat.cast_ofNat, Int.ofNat_toNat,
+      toInt_intCast]
+    rw [Int.max_eq_left]
+    · have h : (2 ^ (w + 1) : Int) = (2 ^ (w + 1) : Nat) := by simp
+      rw [h, Int.emod_bmod]
+    · refine Int.emod_nonneg z ?_
+      exact pow_ne_zero (w + 1) (by decide)
+
+open Fin.CommRing in
+@[simp] theorem toFin_intCast (z : ℤ) : (z : BitVec w).toFin = ↑z := by
+  rw [← ofFin_intCast]
```

Comment:
These are just moved unmodified from further down the file, since they're needed for the smul proofs too.